code/frameworks/svelte-vite/src/plugins/generateDocgen.ts

```diff
@@ -0,0 +1,437 @@
+      types.splice(Math.max(idxTrue, idxFalse), 1);
+      types.splice(Math.min(idxTrue, idxFalse), 1, { type: 'boolean' });
+    }
+
+    return types.length > 1 ? { type: 'union', types: types } : types[0];
+  }
+  if (type.isIntersection()) {
+    // TypeA & TypeB
+    const types = type.types
+      .map((t) => convertType(t, checker))
+      .filter((t) => t !== undefined) as Type[];
+    return { type: 'intersection', types };
+  }
+
+  return undefined;
+}
+
+function initializerToDefaultValue(
+  expr: ts.Expression,
+  checker: ts.TypeChecker
+): DefaultValue | undefined {
+  if (ts.isNumericLiteral(expr)) {
+    return { text: expr.text };
+  } else if (ts.isStringLiteral(expr)) {
+    return { text: JSON.stringify(expr.text) };
+  } else if (ts.isIdentifier(expr)) {
+    const symbol = checker.getSymbolAtLocation(expr);
+    if (symbol) {
+      if (checker.isUndefinedSymbol(symbol)) {
+        return { text: 'undefined' };
+      }
+      const type = checker.getTypeAtLocation(expr);
+      if (type.isLiteral()) {
+        return { text: JSON.stringify(type.value) };
+      } else if (type.flags & ts.TypeFlags.Null) {
+        return { text: 'null' };
+      } else if (type.flags & ts.TypeFlags.BooleanLiteral) {
+        return { text: checker.typeToString(type) };
+      }
+    }
+  }
+  switch (expr.kind) {
+    case ts.SyntaxKind.TrueKeyword:
+      return { text: 'true' };
+    case ts.SyntaxKind.FalseKeyword:
+      return { text: 'false' };
+    case ts.SyntaxKind.NullKeyword:
+      return { text: 'null' };
+  }
+  return undefined;
+}
+
+function loadCompilerOptions(basepath: string): ts.CompilerOptions {
+  const jsConfigPath = ts.findConfigFile(basepath, ts.sys.fileExists, 'jsconfig.json');
+  const tsConfigPath = ts.findConfigFile(basepath, ts.sys.fileExists);
+  const configPath = jsConfigPath || tsConfigPath;
+
+  const forcedOptions = {
+    sourceMap: false,
+    noEmit: true,
+    allowJs: true,
+    checkJs: true,
+    skipLibCheck: true,
+    skipDefaultLibCheck: true,
+  };
+  if (!configPath) {
+    return forcedOptions;
+  }
+
+  const configFile = ts.readConfigFile(configPath, ts.sys.readFile);
+  const config = ts.parseJsonConfigFileContent(
+    configFile.config,
+    ts.sys,
+    path.dirname(configPath),
+    undefined,
+    configPath,
+    undefined
+  );
+
+  return {
+    ...config.options,
+    ...forcedOptions,
+  };
+}
+
+export interface SourceFileCache {
+  filenameToSourceFiles: Record<string, ts.SourceFile>;
+  hashToSourceFiles: Record<string, ts.SourceFile>;
+}
+
+export function createSourceFileCache(): SourceFileCache {
+  return {
+    filenameToSourceFiles: {},
+    hashToSourceFiles: {},
+  };
+}
+
+export function generateDocgen(targetFileName: string, sourceFileCache: SourceFileCache): Docgen {
+  if (targetFileName.endsWith('.svelte')) {
+    targetFileName = targetFileName + '.tsx';
+  }
+  let propsRuneUsed = false;
+
+  const options = loadCompilerOptions(targetFileName);
```

Comment:
Parse tsconfig also scans the file system for the files included in the tsconfig `include` config. It's returned in the `ParsedCommandLine.fileNames`. It might be quite heavy. It might be worth watching the file and cache the result.

The `createProgram` is also heavy so might be better in a larger codebase to reuse the problem. You can probably try creating the program with `ParsedCommandLine.fileNames` but maybe only the svelte files since the ts/js file might never be used. When the `targetFileName` is not in the list or file content changes then recreate the program. It's also worth exploring using the signature with `CreateProgramOptions`. In the option you can pass in the old program to `createProgram` and TypeScript will try to reuse the last program.

Reply:
Thank you very much for your suggestions! I added some caching mechanisms:

- If the previously loaded `ParsedCommandLine.fileNames` includes the target .svelte file, it doesn't reload the tsconfig.
- Passing the previous program as `oldProgram` to createProgram. Subsequent calls to createProgram take less than 10ms for the Storybook `sveltekit/prerelease-ts` example on my machine.

I believe these measures ensure adequate performance. Watching for file changes seems difficult for me and might be too heavy to implement in this PR. What do you think? @jasonlyu123 @JReinhold

Reply:
You can reference the file watching here.
https://github.com/vitejs/vite/blob/8d0a9c1ab8ddd26973509ca230b29604e872e2cd/packages/vite/src/node/plugins/esbuild.ts#L110

One thing I am also curious about is that sometimes the user starts the dev server when svelte-kit hasn't generated the `./svelte-kit/tsconfig.json` file. Does that affect the amount of svelte files being loaded? or maybe the plugin will be loaded after the svelte-kit plugin so it's fine? If it does, maybe the file watch is necessary.